docs/v2/guides/extending-timber.md

```diff
@@ -3,107 +3,471 @@ title: "Extending Timber"
+
+**single-post.twig**
+
+```twig
+<header>
+    <h1>{{ post.title }}</h1>
+    <p>From the {{ post.issue.title }} issue</p>
+</header>
+```
+
+Of course, `Timber\Post` has no built-in concept of an issue. Imagine there’s a custom taxonomy called `issue`. So we're going to extend `Timber\Post` to give it an `issue()` method.
+
+```php
+<?php
+
+class MagazinePost extends \Timber\Post {
+    /**
+     * Issue cache.
+     *
+     * @var \Timber\Term
+     */
+	protected $issue;
+
+    /**
+     * Gets a magazine’s issue.
+     *
+     * @return \Timber|Term|false;
+     */
 	public function issue() {
-		if (!$this->_issue) {
-			$issues = $this->get_terms('issues');
-			if (is_array($issues) && count($issues)) {
-				$this->_issue = $issues[0];
-			}
-		}
-		return $this->_issue;
+        if ( ! empty( $this->_issue ) ) {
+            return $this->issue;
+        }
+
+        $issues = $this->terms( [
+            'taxonomy' => 'issues'
+        ] );
+
+		if ( is_array( $issues ) && ! empty( $issues ) ) {
+			$this->issue = $issues[0];
+        }
+
+        return $this->issue;
 	}
 }
 ```
 
-Right now I'm in the midst of building a complex site for a hybrid foundation and publication. The posts on the site have some very specific requirements that requires a fair amount of logic. I can take the simple `Timber\Post` object and extend it to make it work perfectly for this theme.
+So now we’ve got an easy way to refer to `{{ post.issue }}` in our Twig templates.
+
+### Working with meta fields
 
-For example, I have a plugin that let's people insert manually related posts, but if they don't, WordPress will pull some automatically based on how the post is tagged.
+A very common use case is to use custom methods for handling meta values. Also refer to the [Custom Fields Guide](/docs/v2/guides/custom-fields/) for more information.
+
+Here, we take an custom post type `event` an example.
+
+#### Keep your logic in PHP
+
+An event probably has a date that you define in the meta field `date_start`. You could get the date and manipulate the output format through Twig’s `date` filter.
+
+```twig
+{{ post.meta('date_start')|date('F j Y') }}
+```
+
+But you could also define your own `date_display` method to display that date.
 
 ```php
-	<?php
-	class MySitePost extends Timber\Post {
-
-		function get_related_auto() {
-			$tags = $this->tags();
-			if (is_array($tags) && count($tags)) {
-				$search_tag = $tags[0];
-				$related = new Timber\PostQuery('tag_id='.$search_tag->ID);
-				return $related;
-			} else {
-				//not tagged, cant do related on it
-				return false;
-			}
+/**
+ * Class Event
+ */
+class Event extends \Timber\Post {
+    /**
+	 * Gets display date.
+	 *
+	 * @return string
+	 */
+    public function date_display() {
```

Comment:
I've had to implement this exact use-case a few times now. What I've found to be most powerful is to decouple the date format from the post type completely. Taking an optional parameter here makes that very easy to do:

```
public function date_display( $format = 'F j Y' ) {
  // ...
  return wp_date( $format, $date_start->getTimestamp() );
}
```

This could even apply to the other more complicated start/end date example below, where I imagine you could take `$format_long` and `$format_short` params in place of the hard-coded `F j` and `F j Y` strings, respectively.

I think then providing examples of calling `date_display()` with various formats would illustrate the power of this approach.

Reply:
I make the date formats filterable in my own implementation. Having dynamic date formats is definitely a benefit.

> This could even apply to the other more complicated start/end date example below, where I imagine you could take `$format_long` and `$format_short` params in place of the hard-coded `F j` and `F j Y` strings, respectively.

Yes! Are `F j` and `F j Y` even the right date formats in English? I’m way more used to German date formats.

> I think then providing examples of calling `date_display()` with various formats would illustrate the power of this approach.

Sure! I’m going to add those.

Reply:
@acobster I made some updates in [https://github.com/timber/timber/pull/2260/commits/865d2ddd84ac2eeb32210392880aa4d112b0cf3a ](https://github.com/timber/timber/pull/2260/commits/865d2ddd84ac2eeb32210392880aa4d112b0cf3a). Would you mind checking it again?

Reply:
This looks great! I didn't go over your full date example in great detail (i.e. for logic errors in the example code), but it looks like the right idea. Nice work!

> Are F j and F j Y even the right date formats in English?

Depends on the locale. I think Americans are more used to "January 3rd" whereas a Brit would be more likely to say "3 January"...but probably any English speaker will know what either one means. :)